File: java/algorithm/sort/RadixSort.java
package algorithm.sort;

import java.util.LinkedList;
import java.util.Queue;

/**
 * @author roseduan
 * @time 2019/6/12 22:03
 * @description 基数排序
 */
public class RadixSort {

    private static void radixSort(int[] data) {
        int maxDigit = maxDigit(data);

        //新建并初始化10个桶
        Queue<Integer>[] queues = new LinkedList[10];
        for (int i = 0; i < queues.length; i++) {
            queues[i] = new LinkedList<>();
        }

        for (int i = 0, mod = 1; i < maxDigit; i ++, mod *= 10) {
            for (int n : data){
                int m = (n / mod) % 10;
                queues[m].add(n);
            }

            int count = 0;
            for (Queue<Integer> queue : queues) {
                while (queue.size() > 0) {
                    data[count++] = queue.poll();
                }
            }
        }
    }

    /**
     * 获取数组最大位数
     */
    private static int maxDigit(int[] data){
        int maxDigit = data[0];
        for (int i : data){
            if (maxDigit < i){
                maxDigit = i;
            }
        }
        return String.valueOf(maxDigit).length();
    }
}

File: go/sort/radix_sort.go
package sort

// 基数排序(只适用于整数)
func RadixSort(data []int) {
	if len(data) <= 1 {
		return
	}

	// 找到数组中最大值
	max := data[0]
	for _, v := range data {
		if max < v {
			max = v
		}
	}

	buckets := make([][]int, 10)
	mod := 1
	for ; max > 0; max /= 10 {
		for _, v := range data {
			buckets[(v/mod)%10] = append(buckets[(v/mod)%10], v)
		}

		count := 0
		for i, buc := range buckets {
			for _, v := range buc {
				data[count] = v
				count++
			}
			buckets[i] = []int{}
		}
		mod *= 10
	}
}
